### PR TITLE
Add a lot of debug

### PR DESCRIPTION
**Pull request by bpaquet**

### CODE REVIEW

File: lib/new_relic/agent/pipe_channel_manager.rb [hunk 1]
```diff
@@ -85,12 +97,41 @@ def deserialize_message_length(data)
...
+              @file = File.open('/tmp/marshall_log', 'wb')
+            end
+          end
+          
           @out.close unless @out.closed?
           @in << serialize_message_length(data)
           @in << data
+          if nr_debug?
+            @file << serialize_message_length(data)
+            @file << data
```
Comment (Izzette) [inline lib/new_relic/agent/pipe_channel_manager.rb:118]: Ooo this is a good idea!

File: lib/new_relic/agent/pipe_channel_manager.rb [hunk 2]
```diff
@@ -85,12 +97,41 @@ def deserialize_message_length(data)
...
+          if nr_debug?
+            lock_granted = @write_lock.try_lock
+
+            NewRelic::Agent.logger.error(
+              "Critical code section violated (Thread ID: #{Thread.current.object_id})\n" +
+              "Backtrace:\n" + caller.join("\n").gsub(/^/, "\t")
+            ) unless lock_granted
+
+            unless @file
+              @file = File.open('/tmp/marshall_log', 'wb')
```
Comment (Izzette) [inline lib/new_relic/agent/pipe_channel_manager.rb:109]: Different agents will continually write over this file.  There are multiple pipe channel managers, I believe.  We need to generate either random file names or those unique to the endpoints attached.

File: lib/new_relic/agent/pipe_channel_manager.rb [hunk 3]
```diff
@@ -247,6 +292,11 @@ def unmarshal(data)
         rescue StandardError => e
           ::NewRelic::Agent.logger.error "Failure unmarshalling message from Resque child process, pid #{Process.pid}, size #{data.size}", e
           ::NewRelic::Agent.logger.error Base64.encode64(data)
+          if nr_debug? && File.exist?('/tmp/marshall_log')
+            content = File.binread('/tmp/marshall_log')
+            ::NewRelic::Agent.logger.error "Sent #{content.size}"
```
Comment (Izzette) [inline lib/new_relic/agent/pipe_channel_manager.rb:297]: If this is all we're going to do with our file, we can just open an unnamed inode, if ruby exposes those POSIX APIs that is.  They'll get cleaned up automatically, they'll be synced to disk only lazily as the kernel reclaims VFS cache, we won't have to worry about file name conflicts, etc.

File: lib/new_relic/agent/pipe_channel_manager.rb [hunk 4]
```diff
@@ -247,6 +292,11 @@ def unmarshal(data)
...
+            ::NewRelic::Agent.logger.error Base64.encode64(content)
```
Comment (Izzette) [inline lib/new_relic/agent/pipe_channel_manager.rb:298]: These files might get very very long, I'm not 100% sure.  Maybe we should cap it off at 100 MiB or something?  If it's too big, we won't dump the data.  I think we don't want to screw up staging either.

File: lib/new_relic/agent/pipe_channel_manager.rb (hunk 2 above)
Comment (bpaquet) [reply lib/new_relic/agent/pipe_channel_manager.rb:109]: Nope. In resque we have only one. I tested it on my laptop, seems to works well

File: lib/new_relic/agent/pipe_channel_manager.rb (hunk 3 above)
Comment (bpaquet) [reply lib/new_relic/agent/pipe_channel_manager.rb:297]: I agree this is not really reliable. But I'm seeing datas.

File: lib/new_relic/agent/pipe_channel_manager.rb (hunk 4 above)
Comment (bpaquet) [reply lib/new_relic/agent/pipe_channel_manager.rb:298]: One file per job. I tested it.

**Pull request merged**